vyper/venom/passes/memmerging.py

```diff
@@ -0,0 +1,257 @@
+        # return true if dst overlaps src. this is important for blocking
+        # mcopy batching in certain cases.
+        a = max(self.src_start, self.dst_start)
+        b = min(self.src_end, self.dst_end)
+        return a < b
+
+    def overlap(self, other: "_Interval") -> bool:
+        a = max(self.src_start, other.src_start)
+        b = min(self.src_end, other.src_end)
+        return a < b
+
+    def _add(self, other: "_Interval", ok_dst_overlap: bool) -> bool:
+        if other.src_start != self.src_end:
+            return False
+        if other.dst_start != self.dst_end:
+            return False
+
+        n_inter = _Interval(self.dst_start, self.src_start, self.length + other.length, [])
+        if not ok_dst_overlap and n_inter.dst_overlaps_src():
+            return False
+
+        self.length = n_inter.length
+        self.insts.extend(other.insts)
+        return True
+
+    def merge(self, other: "_Interval", ok_dst_overlap: bool) -> bool:
+        assert self.src_start <= other.src_start, "bad bisect_left"
+        return self._add(other, ok_dst_overlap)
+
+    def __lt__(self, other) -> bool:
+        return self.src_start < other.src_start
+
+    def __repr__(self) -> str:
+        return (
+            f"({self.src_start}, {self.src_end}, {self.length}, {self.dst_start}, {self.dst_end})"
+        )
+
+
+class MemMergePass(IRPass):
+    dfg: DFGAnalysis
+
+    def run_pass(self):
+        self.dfg = self.analyses_cache.request_analysis(DFGAnalysis)  # type: ignore
+
+        for bb in self.function.get_basic_blocks():
+            self._handle_bb_memzero(bb)
+            self._handle_bb(bb, "calldataload", "calldatacopy")
+
+            if version_check(begin="cancun"):
+                # mcopy is available
+                self._handle_bb(bb, "mload", "mcopy")
+
+        self.analyses_cache.invalidate_analysis(DFGAnalysis)
+        self.analyses_cache.invalidate_analysis(LivenessAnalysis)
+
+    def _optimize_copy(self, bb: IRBasicBlock, intervals: list[_Interval], copy_inst: str):
+        for inter in intervals:
+            if inter.length <= 32:
+                continue
+
+            inter.insts[-1].output = None
+            inter.insts[-1].opcode = copy_inst
+            inter.insts[-1].operands = [
+                IRLiteral(inter.length),
+                IRLiteral(inter.src_start),
+                IRLiteral(inter.dst_start),
+            ]
+            for inst in inter.insts[0:-1]:
+                bb.remove_instruction(inst)
+
+        intervals.clear()
+
+    def _add_interval(
+        self, intervals: list[_Interval], new_inter: _Interval, allow_dst_overlap_src: bool = False
+    ) -> bool:
+        if not allow_dst_overlap_src and new_inter.dst_overlaps_src():
+            return False
+        index = bisect_left(intervals, new_inter)
+        intervals.insert(index, new_inter)
+        if index > 0:
+            if intervals[index - 1].overlap(new_inter):
+                return False
```

Comment:
Shouldn't we be removing the interval we just added here? Or do the insert after the checks?